src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java

```diff
@@ -264,12 +266,17 @@ public Proc launch(ProcStarter starter) throws IOException {
                     if (container.isPresent() && container.get().getWorkingDir() != null) {
                         containerWorkingDir = Optional.of(container.get().getWorkingDir());
                     }
-                    if (containerWorkingDir.isPresent() && ! containerWorkingDirFilePath.getRemote().equals(containerWorkingDir.toString())) {
-                        // Container has a custom workingDir, set pwd to the container workspace root
+                    if (containerWorkingDir.isPresent()) {
                         containerWorkingDirStr = containerWorkingDir.get();
-                        containerWorkingDirFilePath = new FilePath(containerWorkingDirFilePath.getChannel(), containerWorkingDirStr);
+                    }
+
+                    if (containerWorkingDir.isPresent() && ! containerWorkingDirFilePath.getRemote().startsWith(containerWorkingDirStr)) {
+                        // Container has a custom workingDir, updated the pwd to match container working dir
+                        containerWorkingDirFilePathStr = containerWorkingDirFilePath.getRemote().replaceFirst(
+                                ContainerTemplate.DEFAULT_WORKING_DIR, containerWorkingDirStr);
```

Comment:
Won't work if the jnlp container uses a custom working directory.

Reply:
@Vlatombe  Setting a custom working directory for JNLP container runs into error much before invoking the kubernetes plugin to execute the shell script. Here is my observation

- Jenkins always cosiders `/home/jenkins/agent` as the workspace directory for declarative pipelines. This is happening because `KubernetesSlave` sets `remoteFs as default working dir in the following method. When this method gets invoked, YAML declaration of the pod was not yet parsed, so, there is no container so, always sets the default container working dir
```java
    public String getRemoteFs() {
        return getFirstContainer().map(ContainerTemplate::getWorkingDir).orElse(ContainerTemplate.DEFAULT_WORKING_DIR);
    }
```
- In the build console output this can be clearly noticed in the following output
`Running on declarative-custom-working-dir-1-0597g-xhtl4-dqklj in /home/jenkins/agent/workspace/declarative Custom Working Dir`
- When executing shell command, Jenkins tried to store the shell commands as a shell script under the remoteFs directory which is `/home/jenkins/agent`. But with in the JNLP container, workspace is mounted in a different directory, not in the `/home/jenkins/agent`. So creation of the shell script in the remote slave fails. This shell script file creation happens with in Jenkins, only after creation of the Shell script control is passed to kubernetes plugin to execute the command.

So, my observation is, for declarative pipelines, JNLP container should always use `/home/jenkins/agent` as working directory. Declaring a different one will result in error. I think the right solution would be to set the `remoteFs` to the working directory of JNLP. But I think that involves a bigger refactoring as YAML has to be parsed prior to setting the `remoteFs`.

I think for now, we can treat JNLP containers not able to use a custom working dir as a limitation. Let me know your thoughts.

Reply:
Looks good to me.

Reply:
> I think for now, we can treat JNLP containers not able to use a custom working dir as a limitation. Let me know your thoughts.

They used to be able to - and indeed when upgrading this broke my pipelines using kaniko...